WordPress/gutenberg
# File: Add experimental integration with Interactivity API

## What?


Explore using [the Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/) for the File block.

It's mostly a follow-up for https://github.com/WordPress/gutenberg/pull/49994 where Interactivity API runtime landed.

I also spend some time looking at how we can scale the Interactivity API experiment so it is easier to convert more blocks using a more unified approach. Part of the effort included the refactoring, where I checked how we could generalize the experiment to all core blocks that want to use Interactivity API.

## Why?


My main goal was to validate the current webpack configuration against integration with the existing core block. I also wanted to see how easy it is to convert an existing block to use Interactivity API.

## How?


I’m still contemplating whether it’s a good idea to convert the File block to Interactivity API, as in practice, it isn’t needed at all… It only increases the bundle size (the size of runtime + directives), which people can use to counterargument the proposal. I see the value in migrating all view scripts to one approach, but I’d love to hear feedback from other folks.

Regarding the block, the only part I wasn't sure about was whether we should hide the PDF preview initially and only enable it when we ensure that the browser can support it. It's the opposite of what is in the `trunk` today, but I also experienced some issues with that approach when I was enforcing Safari to hide the `object` while it started loading the PDF document. It's hard to tell whether it's an actual issue on mobile devices that don't support PDF preview.

I explored a few changes to the webpack config:

- https://github.com/WordPress/gutenberg/pull/50377#discussion_r1187248883
- https://github.com/WordPress/gutenberg/pull/50377#discussion_r1187790587
- https://github.com/WordPress/gutenberg/pull/50377#discussion_r1187793739
- https://github.com/WordPress/gutenberg/pull/50377#discussion_r1187799227

## Testing Instructions





1. Go to the post/page editor.
2. Insert a file block.
3. Upload a PDF file.
4. Ensure that "Show inline embed" option present in the sidebar is checked.
5. Save the post/page.
6. Navigate to the post/page and ensure that the PDF file preview is present in the supported browser.
7. Use a tablet or a mobile device to verify that the PDF preview gets removed.

## Screenshots or screencast 

Show inline embed setting in the editor:

<img width="1466" alt="Screenshot 2023-05-08 at 21 42 46" src="https://user-images.githubusercontent.com/699132/236918093-b71aa44a-e7ef-4041-88a4-49ead0186075.png">


Webpack warning addressed in this PR:

<img width="1303" alt="Screenshot 2023-05-08 at 09 46 06" src="https://user-images.githubusercontent.com/699132/236791243-b9109982-5a5f-4e4e-94fe-c4c452f1db4a.png">


JavaScript size before this PR:

<img width="1245" alt="Screenshot 2023-05-08 at 10 49 10" src="https://user-images.githubusercontent.com/699132/236791259-ebb34a39-1ff5-48c2-bcad-edd53933ee3f.png">

JavaScript size after this PR

<img width="1239" alt="Screenshot 2023-05-08 at 10 44 52" src="https://user-images.githubusercontent.com/699132/236791252-8ee4cab6-ae59-496c-ad68-077af5263422.png">

The size for the File block's view script is nearly the same, but we have to load the Interactivity runtime, which is around 11 kB.
